src/compiler/crystal/semantic/cleanup_transformer.cr

```diff
@@ -249,6 +249,56 @@ module Crystal
       nil
     end
 
+    def transform(node : StringInterpolation)
+      # See if we can solve all the pieces to string literals.
+      # If that's the case, we can replace the entire interpolation
+      # with a single string literal.
+      pieces = node.expressions.dup
+      solve_string_interpolation_expressions(pieces)
+
+      if pieces.all?(StringLiteral)
+        string = pieces.join(&.as(StringLiteral).value)
+        string_literal = StringLiteral.new(string).at(node)
+        string_literal.type = @program.string
+        return string_literal
+      end
+
+      if expanded = node.expanded
+        return expanded.transform(self)
+      end
+      node
+    end
+
+    private def solve_string_interpolation_expressions(pieces : Array(ASTNode))
+      pieces.each_with_index do |piece, i|
+        replacement = solve_string_interpolation_expression(piece)
+        next unless replacement
+
+        pieces[i] = replacement
+      end
+    end
+
+    private def solve_string_interpolation_expression(piece : ASTNode) : StringLiteral?
+      if piece.is_a?(ExpandableNode)
+        expanded = piece.expanded
+        if expanded
```

Comment:
Nit: assignment in the condition to be consistent with the rest of the file (on line 291 as well).

Tangential: Any thoughts about a design guide for Crystal, similar to Dart's? https://dart.dev/guides/language/effective-dart/design

Reply:
There's https://crystal-lang.org/reference/1.5/conventions/coding_style.html but it's probably too small.

Reply:
Wow, the Dart guide is **long**! I wonder how many people read it entirely... I think I prefer something small and memorable.